TesasSemble/graphs.py

```diff
@@ -0,0 +1,164 @@
+
+class Edge:
+    def __init__(self, data, node_a=None, node_b=None):
+        self.data = data
+        self.node_a = node_a
+        self.node_b = node_b
+        if self.node_a is not None and self.node_b is not None:
+            self.node_b.append_in_edges(self)
+            self.node_a.append_out_edges(self)
+
+    def __str__(self):
+        return str(self.data)
+
+    __repr__ = __str__
+
+    def set_node_a(self, node_a):
+        old_a = self.node_a
+        self.node_a = node_a
+        if self.node_b != None:
+            self.node_a.append_out_edges(self)
+            if self not in self.node_b.in_edges:
+                self.node_b.append_in_edges(self)
+            if old_a != None:
+                old_a.remove_out_edges(self)
+
+    def set_node_b(self, node_b):
+        old_b = self.node_b
+        self.node_b = node_b
+        if self.node_a != None:
+            self.node_b.append_in_edges(self)
+            if self not in self.node_a.out_edges:
+                self.node_b.append_out_edges(self)
+            if old_b != None:
+                old_b.remove_in_edges(self)
+
+    def remove_connection(self):
+        if self.node_a is not None and self.node_b is not None:
+            self.node_a.remove_out_edges(self)
+            self.node_b.remove_in_edges(self)
+            self.data = 'EmptyEdge'
+
+    def print_connection(self):
+        if self.node_a is not None and self.node_b is not None:
+            print(self.node_a, '->', self.node_b)
+        elif self.node_a != None:
+            print(self.node_a, '->')
+        elif self.node_b != None:
+            print('->', self.node_b)
+        else:
+            pass
+
+    def get_data(self):
+        return self.data
+
+
+def graph_deep_copy(nodes, edges):
+    '''
+    This function performs a deep copy of a graph. This is made to clone the graph using new instances for
+    Node and Edge. This is useful when modifying the graph temporally.
+
+    Parameters
+    __________
+
+    nodes : dict
+        A dictionary where each key is a str and each value is a Node instance.
+
+    edges : array-like
+        A list of Edge instances that represent the connections among the Node instances in nodes.
+
+    Returns
+    _______
+
+    new_nodes : dict
+        A dictionary where each key is a str and each value is a Node instance. This corresponds to a deep copy of nodes.
+
+    new_edges : array-like
+        A list of Edge instances that represent the connections among the Node instances in new_nodes. This corresponds
+        to a deep copy of edges, but referencing to nodes in new_nodes.
+    '''
+    # Create new instances
+    new_nodes = {}
+    new_edges = []
+
+    # Node instances
+    for node in nodes.values():
+        new_nodes[str(node)] = Node(node.data)
+
+    # Edge instances
+    for edge in edges:
+        new_edges.append(Edge(edge.data, node_a = new_nodes[str(edge.node_a)], node_b = new_nodes[str(edge.node_b)]))
+
+    return new_nodes, new_edges
```

Comment:
new line at the bottom of the file

Reply:
should also tests this with flake8 (will put this in as an issue, can be added to CI also needs and issue)